File: embedded-backend/src/agents/state.js
/**
 * Module containing state classes and data models for the threat designer application.
 * 
 * This module defines the agent state structure and all data models used in threat modeling,
 * converted from Python Pydantic models to JavaScript with Zod schemas for validation.
 */

import { Annotation } from '@langchain/langgraph/web';
import { z } from 'zod';

// ============================================================================
// CONSTANTS
// ============================================================================

const SUMMARY_MAX_WORDS_DEFAULT = 40;
const THREAT_DESCRIPTION_MIN_WORDS = 35;
const THREAT_DESCRIPTION_MAX_WORDS = 50;
const MITIGATION_MIN_ITEMS = 2;
const MITIGATION_MAX_ITEMS = 5;

const AssetType = {
  ASSET: 'Asset',
  ENTITY: 'Entity'
};

const StrideCategory = {
  SPOOFING: 'Spoofing',
  TAMPERING: 'Tampering',
  REPUDIATION: 'Repudiation',
  INFORMATION_DISCLOSURE: 'Information Disclosure',
  DENIAL_OF_SERVICE: 'Denial of Service',
  ELEVATION_OF_PRIVILEGE: 'Elevation of Privilege'
};

const LikelihoodLevel = {
  LOW: 'Low',
  MEDIUM: 'Medium',
  HIGH: 'High'
};

// ============================================================================
// ZOD SCHEMAS FOR DATA MODELS
// ============================================================================

/**
 * Schema for summary state
 */
export const SummaryStateSchema = z.object({
  summary: z.string().describe(`A short headline summary of max ${SUMMARY_MAX_WORDS_DEFAULT} words`)
});

/**
 * Schema for system assets or entities
 */
export const AssetsSchema = z.object({
  type: z.enum([AssetType.ASSET, AssetType.ENTITY])
    .describe(`Type, one of ${AssetType.ASSET} or ${AssetType.ENTITY}`),
  name: z.string().describe('The name of the asset'),
  description: z.string().describe('The description of the asset or entity')
});

/**
 * Schema for collection of assets
 */
export const AssetsListSchema = z.object({
  assets: z.array(AssetsSchema).describe('The list of assets')
});

/**
 * Schema for data flow between entities
 */
export const DataFlowSchema = z.object({
  flow_description: z.string().describe('The description of the data flow'),
  source_entity: z.string().describe('The source entity/asset of the data flow'),
  target_entity: z.string().describe('The target entity/asset of the data flow')
});

/**
 * Schema for trust boundaries
 */
export const TrustBoundarySchema = z.object({
  purpose: z.string().describe('The purpose of the trust boundary'),
  source_entity: z.string().describe('The source entity/asset of the trust boundary'),
  target_entity: z.string().describe('The target entity/asset of the trust boundary')
});

/**
 * Schema for threat sources
 */
export const ThreatSourceSchema = z.object({
  category: z.string().describe('Actor Category'),
  description: z.string().describe('One sentence describing their relevance to this architecture'),
  example: z.string().describe('Brief list of 1-2 specific actor types')
});

/**
 * Schema for collection of flows, boundaries, and threat sources
 */
export const FlowsListSchema = z.object({
  data_flows: z.array(DataFlowSchema).describe('The list of data flows'),
  trust_boundaries: z.array(TrustBoundarySchema).describe('The list of trust boundaries'),
  threat_sources: z.array(ThreatSourceSchema).describe('The list of threat actors')
});

/**
 * Schema for individual threat
 */
export const ThreatSchema = z.object({
  name: z.string().describe(
    'A concise, descriptive name for the threat that clearly identifies the security concern'
  ),
  stride_category: z.enum([
    StrideCategory.SPOOFING,
    StrideCategory.TAMPERING,
    StrideCategory.REPUDIATION,
    StrideCategory.INFORMATION_DISCLOSURE,
    StrideCategory.DENIAL_OF_SERVICE,
    StrideCategory.ELEVATION_OF_PRIVILEGE
  ]).describe(
    `The STRIDE category classification: One of ${Object.values(StrideCategory).join(', ')}`
  ),
  description: z.string().describe(
    `A comprehensive description of the threat scenario, including how it could be executed and its potential consequences. ` +
    `Must be between ${THREAT_DESCRIPTION_MIN_WORDS} and ${THREAT_DESCRIPTION_MAX_WORDS} words. ` +
    `Follow threat grammar: [Threat Actor] + [Action] + [Asset/Target] + [Negative Outcome]`
  ),
  target: z.string().describe(
    'The specific asset, component, system, or data element that could be compromised by this threat'
  ),
  impact: z.string().describe(
    'The potential business, technical, or operational consequences if this threat is successfully exploited. ' +
    'Consider confidentiality, integrity, and availability impacts'
  ),
  likelihood: z.enum([LikelihoodLevel.LOW, LikelihoodLevel.MEDIUM, LikelihoodLevel.HIGH]).describe(
    'The probability of threat occurrence based on factors like attacker motivation, capability, opportunity, and existing controls'
  ),
  mitigations: z.array(z.string())
    .min(MITIGATION_MIN_ITEMS)
    .max(MITIGATION_MAX_ITEMS)
    .describe(
      'Specific security controls, countermeasures, or design changes that can prevent, detect, or reduce the impact of this threat'
    ),
  source: z.string().describe(
    'The threat actor or agent who could execute this threat'
  ),
  prerequisites: z.array(z.string()).default([]).describe(
    'Required conditions, access levels, knowledge, or system states that must exist for this threat to be viable'
  ),
  vector: z.string().describe(
    'The attack vector or pathway through which the threat could be delivered or executed'
  ),
  starred: z.boolean().default(false).optional().describe(
    'User-defined flag for prioritization or tracking. Ignored by automated threat modeling agents'
  )
});

/**
 * Schema for collection of threats
 */
export const ThreatsListSchema = z.object({
  threats: z.array(ThreatSchema).describe('The list of threats')
});

/**
 * Schema for gap analysis continuation decision
 */
export const ContinueThreatModelingSchema = z.object({
  stop: z.boolean().describe(
    'Should continue evaluation further threats or the catalog is comprehensive and complete.'
  ),
  gap: z.string().optional().default('').describe(
    'An in depth gap analysis on how to improve the threat catalog. Required only when \'stop\' is False'
  )
});

// ============================================================================
// DATA MODEL CLASSES
// ============================================================================

/**
 * Model representing the summary of a threat catalog
 */
export class SummaryState {
  constructor(data) {
    const validated = SummaryStateSchema.parse(data);
    Object.assign(this, validated);
  }
}

/**
 * Model representing system assets or entities in threat modeling
 */
export class Assets {
  constructor(data) {
    const validated = AssetsSchema.parse(data);
    Object.assign(this, validated);
  }
}

/**
 * Collection of system assets for threat modeling
 */
export class AssetsList {
  constructor(data) {
    const validated = AssetsListSchema.parse(data);
    Object.assign(this, validated);
  }
}

/**
 * Model representing data flow between entities in a system architecture
 */
export class DataFlow {
  constructor(data) {
    const validated = DataFlowSchema.parse(data);
    Object.assign(this, validated);
  }
}

/**
 * Model representing trust boundaries between entities in system architecture
 */
export class TrustBoundary {
  constructor(data) {
    const validated = TrustBoundarySchema.parse(data);
    Object.assign(this, validated);
  }
}

/**
 * Model representing sources of threats in the system
 */
export class ThreatSource {
  constructor(data) {
    const validated = ThreatSourceSchema.parse(data);
    Object.assign(this, validated);
  }
}

/**
 * Collection of data flows, trust boundaries, and threat sources
 */
export class FlowsList {
  constructor(data) {
    const validated = FlowsListSchema.parse(data);
    Object.assign(this, validated);
  }
}

/**
 * Model representing an identified security threat using the STRIDE methodology
 */
export class Threat {
  constructor(data) {
    const validated = ThreatSchema.parse(data);
    Object.assign(this, validated);
  }
}

/**
 * Collection of identified security threats
 */
export class ThreatsList {
  constructor(data) {
    const validated = ThreatsListSchema.parse(data);
    Object.assign(this, validated);
  }

  /**
   * Combine two ThreatsList instances, avoiding duplicates based on name
   * @param {ThreatsList} other - Another ThreatsList to combine with
   * @returns {ThreatsList} Combined threats list
   */
  add(other) {
    const existingNames = new Set(this.threats.map(t => t.name));
    const newThreats = other.threats.filter(t => !existingNames.has(t.name));
    const combinedThreats = [...this.threats, ...newThreats];
    return new ThreatsList({ threats: combinedThreats });
  }
}

/**
 * Tool to share the gap analysis for threat modeling
 */
export class ContinueThreatModeling {
  constructor(data) {
    const validated = ContinueThreatModelingSchema.parse(data);
    Object.assign(this, validated);
  }
}

// ============================================================================
// AGENT STATE DEFINITION
// ============================================================================

/**
 * Custom reducer for threat_list that combines ThreatsList instances
 * @param {ThreatsList} x - Current state
 * @param {ThreatsList} y - New state
 * @returns {ThreatsList} Combined state
 */
function threatListReducer(x, y) {
  if (!x) return y;
  if (!y) return x;
  
  // Merge threat lists, avoiding duplicates
  const existingNames = new Set(x.threats.map(t => t.name));
  const newThreats = y.threats.filter(t => !existingNames.has(t.name));
  
  return {
    threats: [...x.threats, ...newThreats]
  };
}

/**
 * Custom reducer for gap array that appends new gaps
 * @param {Array<string>} x - Current state
 * @param {Array<string>} y - New state
 * @returns {Array<string>} Combined state
 */
function gapReducer(x, y) {
  if (!x) return y || [];
  if (!y) return x;
  return [...x, ...y];
}

/**
 * Default reducer that replaces old value with new value
 * @param {*} x - Current state
 * @param {*} y - New state
 * @returns {*} New state or current state if new is undefined
 */
function defaultReducer(x, y) {
  return y ?? x;
}

/**
 * Container for the internal state of the threat modeling agent
 * 
 * This uses LangGraph's Annotation.Root to define the state structure
 * with appropriate reducers for each field.
 */
export const AgentState = Annotation.Root({
  job_id: Annotation({ reducer: defaultReducer }),
  summary: Annotation({ reducer: defaultReducer }),
  assets: Annotation({ reducer: defaultReducer }),
  image_data: Annotation({ reducer: defaultReducer }),
  system_architecture: Annotation({ reducer: defaultReducer }),
  description: Annotation({ reducer: defaultReducer }),
  assumptions: Annotation({ reducer: defaultReducer }),
  threat_list: Annotation({ reducer: threatListReducer }),
  iteration: Annotation({ reducer: defaultReducer }),
  retry: Annotation({ reducer: defaultReducer }),
  s3_location: Annotation({ reducer: defaultReducer }),
  title: Annotation({ reducer: defaultReducer }),
  owner: Annotation({ reducer: defaultReducer }),
  stop: Annotation({ reducer: defaultReducer }),
  gap: Annotation({ reducer: gapReducer }),
  replay: Annotation({ reducer: defaultReducer }),
  instructions: Annotation({ reducer: defaultReducer })
});

// ============================================================================
// EXPORTS
// ============================================================================

export {
  AssetType,
  StrideCategory,
  LikelihoodLevel,
  SUMMARY_MAX_WORDS_DEFAULT,
  THREAT_DESCRIPTION_MIN_WORDS,
  THREAT_DESCRIPTION_MAX_WORDS,
  MITIGATION_MIN_ITEMS,
  MITIGATION_MAX_ITEMS
};

File: backend/threat_designer/state.py
"""Module containing state classes and data models for the threat designer application."""

import operator
from datetime import datetime
from langgraph.graph import MessagesState
from typing import Annotated, List, Literal, Optional, TypedDict

from constants import (
    CRITICALITY_MULTIPLIER_MAP,
    MITIGATION_MAX_ITEMS,
    MITIGATION_MIN_ITEMS,
    SUMMARY_MAX_WORDS_DEFAULT,
    AssetType,
    CriticalityLevel,
    StrideCategory,
)
from langchain_aws import ChatBedrockConverse
from pydantic import BaseModel, Field


class ConfigSchema(TypedDict):
    """Configuration schema for the workflow."""

    model_assets: ChatBedrockConverse
    model_flows: ChatBedrockConverse
    model_threats: ChatBedrockConverse
    model_gaps: ChatBedrockConverse
    model_struct: ChatBedrockConverse
    model_summary: ChatBedrockConverse
    start_time: datetime
    reasoning: bool


class SummaryState(BaseModel):
    """Model representing the summary of a threat catalog."""

    summary: Annotated[
        str,
        Field(
            description=f"A short headline summary of max {SUMMARY_MAX_WORDS_DEFAULT} words"
        ),
    ]


class Assets(BaseModel):
    """Model representing system assets or entities in threat modeling."""

    type: Annotated[
        Literal[AssetType.ASSET.value, AssetType.ENTITY.value],
        Field(
            description=f"Type, one of {AssetType.ASSET.value} or {AssetType.ENTITY.value}"
        ),
    ]
    name: Annotated[str, Field(description="The name of the asset")]
    description: Annotated[
        str, Field(description="The description of the asset or entity")
    ]
    criticality: Annotated[
        Literal["Low", "Medium", "High"],
        Field(description="Criticality level of the asset", default="Medium"),
    ] = "Medium"

    @property
    def criticality_multiplier(self) -> float:
        """Computed multiplier from criticality level. Not persisted to DynamoDB."""
        return CRITICALITY_MULTIPLIER_MAP.get(self.criticality, 1.0)


class AssetsList(BaseModel):
    """Collection of system assets for threat modeling."""

    assets: Annotated[List[Assets], Field(description="The list of assets")]


class DataFlow(BaseModel):
    """Model representing data flow between entities in a system architecture."""

    flow_description: Annotated[
        str, Field(description="The description of the data flow")
    ]
    source_entity: Annotated[
        str, Field(description="The source entity/asset of the data flow")
    ]
    target_entity: Annotated[
        str, Field(description="The target entity/asset of the data flow")
    ]


class TrustBoundary(BaseModel):
    """Model representing trust boundaries between entities in system architecture."""

    purpose: Annotated[str, Field(description="The purpose of the trust boundary")]
    source_entity: Annotated[
        str, Field(description="The source entity/asset of the trust boundary")
    ]
    target_entity: Annotated[
        str, Field(description="The target entity/asset of the trust boundary")
    ]


class ContinueThreatModeling(BaseModel):
    """Tool to share the gap analysis for threat modeling."""

    stop: Annotated[
        bool,
        Field(
            description="Should stop threat generation (True) if catalog is comprehensive, or continue (False) if gaps remain."
        ),
    ]
    gap: Annotated[
        Optional[str],
        Field(
            description="""Concise list of identified gaps to improve the threat model. \n
            Each gap must be maximum 40 words. \n
            Format as a bulleted list with each gap on a new line. \n
            Gaps should be actionable and specific. \n
            Very important! When referencing threats, do not use abbreviations. Instead you need to reference the full name of the threat, which is their unique identifier. \n
            Required only when 'stop' is False. \n"""
        ),
    ] = ""
    rating: Annotated[
        int,
        Field(
            description="Overall quality rating of the threat catalog (1-10). 10 = comprehensive and high quality, 1 = significant gaps and issues.",
            ge=1,
            le=10,
        ),
    ]


class ThreatSource(BaseModel):
    """Model representing sources of threats in the system."""

    category: Annotated[str, Field(description="Actor Category")]
    description: Annotated[
        str,
        Field(
            description="One sentence describing their relevance to this architecture"
        ),
    ]
    example: Annotated[str, Field(description="Brief list of 1-2 specific actor types")]


class FlowsList(BaseModel):
    """Collection of data flows, trust boundaries, and threat sources."""

    data_flows: Annotated[List[DataFlow], Field(description="The list of data flows")]
    trust_boundaries: Annotated[
        List[TrustBoundary], Field(description="The list of trust boundaries")
    ]
    threat_sources: Annotated[
        List[ThreatSource], Field(description="The list of threat actors")
    ]


class Threat(BaseModel):
    """Model representing an identified security threat using the STRIDE methodology."""

    name: Annotated[
        str,
        Field(
            description="A concise, descriptive name for the threat that clearly identifies the security concern"
        ),
    ]
    stride_category: Annotated[
        Literal[*[category.value for category in StrideCategory]],
        Field(
            description=f"The STRIDE category classification: One of {', '.join([category.value for category in StrideCategory])}."
        ),
    ]
    description: Annotated[
        str,
        Field(
            description="Threat description which must follow threat grammar template format:"
            "[threat source] [prerequisites] can [threat action] which leads to [threat impact], negatively impacting [impacted assets]."
        ),
    ]
    target: Annotated[
        str,
        Field(
            description="The specific asset, component, system, or data element that could be compromised by this threat. It must be only one entity, multiple ones will result in rejecting the threat"
        ),
    ]
    impact: Annotated[
        str,
        Field(
            description="The potential business, technical, or operational consequences if this threat is successfully exploited. Consider confidentiality, integrity, and availability impacts"
        ),
    ]
    likelihood: Annotated[
        Literal["Low", "Medium", "High"],
        Field(
            description="The probability of threat occurrence based on factors like attacker motivation, capability, opportunity, and existing controls"
        ),
    ]
    mitigations: Annotated[
        List[str],
        Field(
            description="Specific security controls, countermeasures, or design changes that can prevent, detect, or reduce the impact of this threat",
            min_items=MITIGATION_MIN_ITEMS,
            max_items=MITIGATION_MAX_ITEMS,
        ),
    ]
    source: Annotated[
        str,
        Field(
            description="The threat actor or agent who could execute this threat. Must match with the category field in threat_source ",
        ),
    ]
    prerequisites: Annotated[
        List[str],
        Field(
            description="Required conditions, access levels, knowledge, or system states that must exist for this threat to be viable",
        ),
    ]
    vector: Annotated[
        str,
        Field(
            description="The attack vector or pathway through which the threat could be delivered or executed",
        ),
    ]
    starred: Annotated[
        bool,
        Field(
            description="User-defined flag for prioritization or tracking. Ignored by automated threat modeling agents",
        ),
    ] = False
    notes: Annotated[
        Optional[str],
        Field(
            description="Reserved for user annotations. Do not read or write this field.",
            default=None,
        ),
    ] = None


class ThreatsList(BaseModel):
    """Collection of identified security threats."""

    threats: Annotated[List[Threat], Field(description="The list of threats")]

    def __add__(self, other: "ThreatsList") -> "ThreatsList":
        """Combine two ThreatsList instances, avoiding duplicates based on name."""
        existing_names = {threat.name for threat in self.threats}
        new_threats = [
            threat for threat in other.threats if threat.name not in existing_names
        ]
        combined_threats = self.threats + new_threats
        return ThreatsList(threats=combined_threats)

    def remove(self, threat_name: str) -> "ThreatsList":
        """Remove a threat by name and return a new ThreatsList instance."""
        filtered_threats = [
            threat for threat in self.threats if threat.name != threat_name
        ]
        return ThreatsList(threats=filtered_threats)


def create_constrained_threat_model(
    asset_names: set[str], source_categories: set[str]
) -> tuple[type[BaseModel], type[BaseModel]]:
    """Create Threat and ThreatsList models with Literal-constrained target/source fields.

    Uses pydantic.create_model() to dynamically override the `target` and `source`
    fields with Literal types when valid values are available, falling back to str
    when the respective set is empty.

    Args:
        asset_names: Set of valid asset names from state.assets.
        source_categories: Set of valid threat source categories.

    Returns:
        Tuple of (DynamicThreat, DynamicThreatsList) model classes.
    """
    from pydantic import create_model

    field_overrides = {}

    if asset_names:
        target_literal = Literal[tuple(sorted(asset_names))]
        field_overrides["target"] = (
            Annotated[
                target_literal,
                Field(
                    description="The specific asset that could be compromised by this threat. Must exactly match one of the allowed values."
                ),
            ],
            ...,
        )

    if source_categories:
        source_literal = Literal[tuple(sorted(source_categories))]
        field_overrides["source"] = (
            Annotated[
                source_literal,
                Field(
                    description="The threat actor category. Must exactly match one of the allowed values."
                ),
            ],
            ...,
        )

    DynamicThreat = create_model("Threat", __base__=Threat, **field_overrides)

    DynamicThreatsList = create_model(
        "ThreatsList",
        __base__=ThreatsList,
        threats=(
            Annotated[List[DynamicThreat], Field(description="The list of threats")],
            ...,
        ),
    )

    return DynamicThreat, DynamicThreatsList


class AgentState(TypedDict):
    """Container for the internal state of the threat modeling agent."""

    summary: Optional[str] = None
    assets: Optional[AssetsList] = None
    image_data: Optional[str] = None
    image_type: Optional[str] = None
    system_architecture: Optional[FlowsList] = None
    description: Optional[str] = None
    assumptions: Optional[List[str]] = None
    improvement: Optional[str] = None
    next_step: Optional[str] = None
    threat_list: Annotated[ThreatsList, operator.add]
    job_id: Optional[str] = None
    retry: Optional[int] = 1
    iteration: Optional[int] = 0
    s3_location: Optional[str]
    title: Optional[str] = None
    owner: Optional[str] = None
    stop: Optional[bool] = False
    replay: Optional[bool] = False
    instructions: Optional[str] = None
    application_type: Optional[str] = "hybrid"


def _add_or_overwrite(left, right):
    """Reducer that adds deltas or replaces with Overwrite.

    Tools send deltas (+1, +2) for increments, or Overwrite(value) for resets.
    """
    from langgraph.types import Overwrite

    # If right is Overwrite, use its value (explicit replacement like reset to 0)
    if isinstance(right, Overwrite):
        return right.value
    # If left is Overwrite, use right
    if isinstance(left, Overwrite):
        return right
    # Otherwise add the delta
    return left + right


def _overwrite_or_last(left, right):
    """Reducer for boolean flags that respects Overwrite."""
    from langgraph.types import Overwrite

    if isinstance(right, Overwrite):
        return right.value
    if isinstance(left, Overwrite):
        return right
    return right


class ThreatState(MessagesState):
    """Container for the internal state of the subgraph."""

    threat_list: Annotated[ThreatsList, operator.add]
    tool_use: Annotated[int, _add_or_overwrite] = 0
    gap_tool_use: Annotated[int, _add_or_overwrite] = 0
    assets: Optional[AssetsList] = None
    image_data: Optional[str] = None
    image_type: Optional[str] = None
    system_architecture: Optional[FlowsList] = None
    description: Optional[str] = None
    assumptions: Optional[List[str]] = None
    gap: Annotated[List[str], operator.add] = []
    instructions: Optional[str] = None
    job_id: Optional[str] = None
    retry: Optional[int] = 1
    iteration: Optional[int] = 0
    replay: Optional[bool] = False
    application_type: Optional[str] = "hybrid"
